File: LinkedListCycleII.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None
        
class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None or head.next == None:
            return None
        fast = head
        slow = head
        while fast.next!=None and fast.next.next!=None:
            fast = fast.next.next
            slow = slow.next
            if fast == slow:
                begin = head
                while begin!=slow:
                    begin = begin.next
                    slow = slow.next
                return begin
                break;
        return None
        
if __name__=='__main__':
    a = ListNode(1);
    b = ListNode(2);
    c = ListNode(3);
    d = ListNode(4);
    e = ListNode(5);
    
    a.next = b;
    b.next = c;
    c.next = d;
    d.next = e;
    e.next = e;
    
    solution = Solution()
    print(solution.detectCycle(a).val)

File: LinkedListCycleII.cpp
#include<iostream>
#include<unordered_map>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        
        unordered_map<ListNode*, int> map;
        ListNode *p = head;
        while(p){
            if(map[p]) return p;
            else map[p] = 1;
            p = p->next;
        }
        
        return NULL;
        
    }
};

int main(){
    ListNode *a = new ListNode(1);
    ListNode *b = new ListNode(2);
    ListNode *c = new ListNode(3);
    ListNode *d = new ListNode(4);
    ListNode *e = new ListNode(5);
    
    a->next = b;
    b->next = c;
    c->next = d;
    d->next = e;
    e->next = a;
    
    Solution *solution = new Solution();
    ListNode *begin = solution->detectCycle(a);
    cout<<begin->val;
    
    return 0;
}
